File: 0273_Modest - 2013 - Radiative heat transfer/代码/第三版/chapter_02_03/dirreflec.m
function dirreflec()

A = input('INPUT INDEX OF REFRACTION: [N, K]:  ');
AN = A(1); 
AK = A(2);
fprintf('N = %f, K = %f \n', AN, AK);
ISTEP = input('SPACING FOR OUTPUT, IN DEGREES:   ');


fprintf('     THETA     RHOS      RHOP      RHO       EPS \n');
PI2=2.E0*atan(1.E0);
j = 1;
for I = 1:ISTEP:90
    TH(j) = (I-1)*PI2/9.E1;
    PH(j) = (I-1);
    [RHOS(j),RHOP(j),RHO(j)] = fresnel(AN,AK,TH(j));
    EPS(j) = 1.E0 - RHO(j);
    fprintf(' %g %g %g %g %g \n' , ...
            PH(j), RHOS(j), RHOP(j),RHO(j), EPS(j));
    j = j + 1;
end
h = polar(repmat(TH',1,4), [RHOS', RHOP', RHO', EPS']);
set(h, 'LineWidth', 2);
legend('RHOS', 'RHOP', 'RHO', 'EPS');

function [RHOS, RHOP, RHO] = fresnel(AN, AK, TH)  

C1 = AN^2 - AK^2 - (sin(TH))^2;
P2 = .5E0*(sqrt(C1^2 + (2.E0*AN*AK)^2) + C1);
Q2 = P2 - C1;
CT = cos(TH);
C2 = (AN^2-AK^2)*CT;
C3 = 2.E0*AN*AK*CT;
P  = sqrt(P2);
Q  = sqrt(Q2);
RHOP = ((P-C2)^2 + (Q-C3)^2)/((P+C2)^2 + (Q+C3)^2);
RHOS = ((CT-P)^2 + Q2)/((CT+P)^2 + Q2);
RHO  =  .5E0*(RHOP + RHOS);

File: 0273_Modest - 2013 - Radiative heat transfer/代码/第三版/chapter_02_03/dirreflec.cpp
// Dirreflec

# include <iostream.h>
# include <math.h>
# include <iomanip.h>

void FRESNEL( double, double, double, double&, double&, double&);

int main()
	{

	double AN,AK,ISTEP,PI2,TH,PH,EPS,I,RHOS,RHOP,RHO;

	cout<<"     INPUT INDEX OF REFRACTION: N,K"<<endl;
	cin>>AN>>AK;
	cout<<"     N = "<<AN<<",  K = "<<AK<<endl;
	cout<<"     INPUT SPACING FOR OUTPUT, IN DEGREES"<<endl;
	cin>>ISTEP;
	cout<<"     THETA     RHOS      RHOP      RHO       EPS"<<endl;
	PI2=2.E0*atan(1.E0);
	for(I=1;I<91;I=I+ISTEP)
		{
		TH=(I-1)*PI2/9.E1;
		PH=(I-1);
		FRESNEL(AN,AK,TH,RHOS,RHOP,RHO);
		EPS=1.E0-RHO;
		cout<<setw(8)<<PH<<setw(13)<<RHOS<<setw(10)<<RHOP<<setw(10)<<
		RHO<<setw(10)<<EPS<<endl;
		}
	return 0;
	}


void FRESNEL(double AN, double AK, double TH, double& RHOS, double& RHOP,
				double& RHO)
{
	double C1,P2,Q2,CT,C2,C3,P,Q;

	C1=AN*AN-AK*AK-pow(sin(TH),2);
	P2=.5E0*(sqrt(C1*C1+pow((2.E0*AN*AK),2))+C1);
	Q2=P2-C1;
	CT=cos(TH);
	C2=(AN*AN-AK*AK)*CT;
	C3=2.E0*AN*AK*CT;
	P=sqrt(P2);
	Q=sqrt(Q2);
	RHOP=(pow((P-C2),2)+pow((Q-C3),2))/(pow((P+C2),2)+pow((Q+C3),2));
	RHOS=(pow((CT-P),2)+Q2)/(pow((CT+P),2)+Q2);
	RHO=.5E0*(RHOP+RHOS);
}


